keps/20200901-standard-crd.md

```diff
@@ -0,0 +1,254 @@
+
+### Stores
+
+These stores are relevant:
+* AWS Secure Systems Manager Parameter Store
+* AWS Secrets Manager
+* Hashicorp Vault
+* Azure Key Vault
+* Alibaba Cloud KMS Secret Manager
+* Google Cloud Platform Secret Manager
+* Kubernetes (see #422)
+* noop (see #476)
+
+### Frontends
+
+* Kind=Secret
+* *potentially* we could sync store to store
+
+## Proposal
+
+### API
+
+### External Secret
+
+The `ExternalSecret` CustomResourceDefinition is **namespaced**. It defines the following:
+1. source for the secret (store)
+2. sink for the secret (fronted)
+3. and a mapping to translate the keys
+
+```yaml
+apiVersion: external-secrets.k8s.io/v1alpha1
+kind: ExternalSecret
+metadata: {...}
+
+spec:
+
+  # the amount of time before the values will be read again from the store
+  refreshInterval: "1h"
+
+  # there can only be one target per ES
+  target:
+    # The secret name of the resource
+    # defaults to .metadata.name of the ExternalSecret. immutable.
+    name: my-secret
+
+    # Enum with values: 'Owner', 'Merge', or 'None'
+    # Default value of 'Owner'
+    # Owner creates the secret and sets .metadata.ownerReferences of the resource
+    # Merge does not create the secret, but merges in the data fields to the secret
+    # None does not create a secret (future use with injector)
+    creationPolicy: 'Merge'
+
+    # specify a blueprint for the resulting Kind=Secret
+    template:
+      type: kubernetes.io/dockerconfigjson # or TLS...
+
+      metadata:
+        annotations: {}
+        labels: {}
+
+      # use inline templates to construct your desired config file that contains your secret
+      data:
+        config.yml: |
+          endpoints:
+          - https://{{ .data.user }}:{{ .data.password }}@api.exmaple.com
+
+      # Uses an existing template from configmap
+      # secret is fetched, merged and templated within the referenced configMap data
+      # It does not update the configmap, it creates a secret with: data["alertmanager.yml"] = ...result...
+      templateFrom:
+      - configMap:
+          name: alertmanager
+          items:
+          - key: alertmanager.yaml
+
+  # data contains key/value pairs which correspond to the keys in the resulting secret
+  data:
+
+    # EXAMPLE 1: simple mapping
+    # one key from a store may hold multiple values
+    # we need a way to map the values to the frontend
+    # it is the responsibility of the store implementation to know how to extract a value
+    tls.crt:
+      key: /corp.org/dev/certs/ingress
+      property: pubcert
+    tls.key:
+      key: /corp.org/dev/certs/ingress
+      property: privkey
+
+  # used to fetch all properties from a secret.
+  # if multiple dataFrom are specified, secrets are merged in the specified order
+  dataFrom:
+  - key: /user/all-creds
+
+# status holds the timestamp and status of last last sync
+status:
+  lastSync: 2020-09-01T18:19:17.263Z # ISO 8601 date string
+  status: success # or failure
+
```

Comment:
A `message` field with the stringified error would be great for diagnostics.

Reply:
Agree, any reason not to go with a condition array block here for readiness?

Reply:
Sure, conditions would fit in here nicely!

What about:
```yaml
status:
  # represents the current phase of secret sync:
  # * Pending | ES created, controller did not (yet) sync the ES or other dependencies are missing (e.g. secret store or configmap template)
  # * Syncing | ES is being actively synced according to spec
  # * Failing | Secret can not be synced, this might require user intervention
  # * Failed  | ES can not be synced right now and will not able to
  # * Completed | ES was synced successfully (one-time use only)
  phase: Syncing
  lastSyncTime: 2020-09-01T18:19:17.263Z # ISO 8601 date string
  failureReason: "..."
  failureMessage: "..."
  
  conditions:
  - type: InSync
    status: "True" # False if last sync was not successful
    reason: "SecretSynced"
    message: "Secret was synced"
    lastTransitionTime: "2019-08-12T12:33:02Z"
    lastSyncTime: "2020-09-23T16:27:53Z"
```

Reply:
Looks good, a few comments
* `lastHeartbeatTime` doesnt need to be used here, that is generally only for conditions which need explicit heart beats
* `failureReason`, `failureMessage` already fit within the condition, those would better fit with events in this case rather than explicitly in the base status.

Reply:
Regarding `lastHeartbeatTime`: i think this is the case here. The `InSync` condition should be updated with every sync. We should consider renaming this to `lastSyncTime`.

`failure*` as attributes are duplicates, true. Ill remove them.